src/coreclr/System.Private.CoreLib/src/System/Attribute.CoreCLR.cs

```diff
@@ -438,6 +438,9 @@ private static Attribute[] CreateAttributeArrayHelper(Type elementType, int elem
         {
             return (Attribute[])Array.CreateInstance(elementType, elementCount);
         }
+
+        private static Attribute[] ToAttributeArray(object[] getCustomAttributesResult) =>
+            getCustomAttributesResult as Attribute[] ?? Array.Empty<Attribute>();
```

Comment:
This was also an option I considered, but I wasn't sure if there's a code path that would return a populated array (somehow) and now those elements would be discarded.

Reply:
I agree with @Joe4evr, a null value only returned when a generic attribute requested, adding null check for all overlaods wold affect perf. I think the the code section @Joe4evr [pointed out in his comment](https://github.com/dotnet/runtime/issues/64335#issuecomment-1026193422) is where we should lookup for a fix
> We could fall back to an empty Attribute[] (what I went with here).

That is the expected behavior when there is no any attribute found

Reply:
@buyaa-n can you clarify what you're suggesting here?

> I think the the code section @Joe4evr https://github.com/dotnet/runtime/issues/64335#issuecomment-1026193422 is where we should lookup for a fix

I did change those lines (see below). However, that doesn't solve the general problem that having an `as` cast here will result in returning null for any case where the underlying `GetCustomAttributes()` function doesn't return an `Attribute[]`. It didn't feel right to leave that in place.

> adding null check for all overlaods wold affect perf

Are you saying that adding `?? Array.Empty<Attribute>()` to handle the "cast failed" case will be a perf issue? If so, how do you think we should handle the case where the cast fails?

> I wasn't sure if there's a code path that would return a populated array (somehow) and now those elements would be discarded.

As I understand it, what @Joe4evr is saying here is that we could end up handling values like `new object[] { new SomeAttribute() }` or `new string[] { "a" }` or even `new object[] { new SomeAttribute(), "a" }`. What is the preferred behavior in such cases? Before this change we would return `null`. After this change we will return `Array.Empty<Attribute>()`. However, we could do something else.

Reply:
> I did change those lines (see below). However, that doesn't solve the general problem that having an as cast here will result in returning null for any case where the underlying GetCustomAttributes() function doesn't return an Attribute[]. It didn't feel right to leave that in place.

I mean it was not returning (and not expected to return) null before this open generic attributes case introduced, and as you fixed that open generic case below i don't think we need this defensive mechanism, even if by the existing code it could return null in case an object array returned from the internal calls. Though I understand your reasoning, but as @Joe4evr mentioned now that object array creation will be discarded, probably we should change the object array returning logic instead

Reply:
> i don't think we need this defensive mechanism, even if by the existing code it could return null in case an object array returned from the internal calls

Ok thanks I understand. Do you think we should return to the `as` cast from before or should we do a "strict" cast `(Attribute[])member.GetCustomAttributes(...)`?

Also, I'm curious to get your thoughts on the desired behavior for the `CustomFieldInfo` test case I added:
| Case | `memberInfo.GetCustomAttributes()` return value | Old behavior | Initial PR behavior |
| ---- | ---- | ---- | ---- |
| 1 | `null` | `return null` | `return Array.Empty<Attribute>()` |
| 2 | `new object[0]` |  `return null` | `return Array.Empty<Attribute>()` |
| 3 | `new string[] { "a" }` | `return null` | `return Array.Empty<Attribute>()` |
| 4 | `new object[] { new GenericAttribute<string>, "b" }` | `return null` | `return Array.Empty<Attribute>()` |

> now that object array creation will be discarded, probably we should change the object array returning logic instead

As I understand it, this comment is referring to making further changes in the `CustomAttribute` class in `RuntimeCustomAttributeData.cs` to stop returning `object[]` at all. Currently, the cases where we still return an array with a non-attribute element type are:
(1) The passed in "attribute" type is a value type
(2) The passed in "attribute" type is a reference type that is not assignable to `Attribute`.

For example, here's some code that shows how this behavior is externally visible:
```C#
        Console.WriteLine(typeof(int).GetCustomAttributes(typeof(string), false).GetType()); // System.String[]
	Console.WriteLine(typeof(int).GetCustomAttributes(typeof(List<>), false).GetType()); // System.Object[]
	Console.WriteLine(typeof(int).GetCustomAttributes(typeof(int), false).GetType()); // System.Object[]
	Console.WriteLine(typeof(int).GetCustomAttributes(typeof(object), false).GetType()); // System.Object[]
	Console.WriteLine(typeof(int).GetCustomAttributes(typeof(object), false).Length); // 3 (actually populated Object[]!)

        // also works with TypeBuilder
	var assembly = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("asm"), AssemblyBuilderAccess.RunAndCollect);
	var module = assembly.DefineDynamicModule("mod");
	var type = module.DefineType("t");
	type.CreateType();
	Console.WriteLine(type.GetCustomAttributes(typeof(string), false).GetType()); // System.String[]
	Console.WriteLine(type.GetCustomAttributes(typeof(List<>), false).GetType()); // System.Object[]
	Console.WriteLine(type.GetCustomAttributes(typeof(int), false).GetType()); // System.Object[]
```

The reason that this behavior leaks out is that `RuntimeType` and `TypeBuilder` call into `CustomAttribute` directly without validating that the passed in type is an `Attribute` type for this one overload of `GetCustomAttributes`. This frankly feels like a bug to me, although the ability to request `object` does actually work.

Thoughts on what the best path forward is here and what impliciations there are for how we should handle unexpected return values in `Attribute.cs`?

Reply:
Note that it is fine for attributes to not inherit from `System.Attribute`. It is not possible to create such attributes in C#, but it is possible in IL and other languages, in addition to the Reflection.Emit case that you have run into.

Reply:
@jkotas which framework APIs are supposed to work with such attributes?

Today, it seems that most do not work because you run into a cast to Attribute somewhere along the way.

Reply:
APIs that return `object` and `object[]`: `MemberInfo.GetCustomAttributes`, etc.

Reply:
> Ok thanks I understand. Do you think we should return to the as cast from before or should we do a "strict" cast (Attribute[])member.GetCustomAttributes(...)?

Personally I would use strict casting and let it throw, and it will be same as `GetCustomAttributes` for `Assemply` which throws InvalidCastException in such situation